runtime/runtime/src/lib.rs

```diff
@@ -1294,62 +1297,125 @@ impl Runtime {
-                    Some((account_id.clone(), code.len() as u64))
-                }
-                StateRecord::AccessKey { account_id, public_key, access_key } => {
-                    let public_key: PublicKey = public_key.clone();
-                    let access_key: AccessKey = access_key.clone().into();
-                    let storage_usage = config.num_extra_bytes_record
-                        + public_key.try_to_vec().unwrap().len() as u64
-                        + access_key.try_to_vec().unwrap().len() as u64;
-                    Some((account_id.clone(), storage_usage))
-                }
-                StateRecord::PostponedReceipt(_) => None,
-                StateRecord::ReceivedData { .. } => None,
-                StateRecord::DelayedReceipt(_) => None,
-            };
-            if let Some((account, storage_usage)) = account_and_storage {
-                *result.entry(account).or_default() += storage_usage;
-            }
-        }
-        result
+        let mut storage_computer = StorageComputer::new(config);
+        storage_computer.process_records(records);
+        storage_computer.finalize()
     }
+}
 
-    /// Balances are account, publickey, initial_balance, initial_tx_stake
-    pub fn apply_genesis_state<Record: Borrow<StateRecord>>(
-        &self,
+struct GenesisStateApplier<'a> {
+    tries: ShardTries,
+    shard_id: ShardId,
+    validators: &'a [(AccountId, PublicKey, Balance)],
+    config: &'a RuntimeConfig,
+    current_state_root: CryptoHash,
+    delayed_receipts_indices: DelayedReceiptIndices,
+}
+
+impl<'a> GenesisStateApplier<'a> {
+    pub fn new(
         tries: ShardTries,
         shard_id: ShardId,
-        validators: &[(AccountId, PublicKey, Balance)],
-        records: &[Record],
-        config: &RuntimeConfig,
-    ) -> (StoreUpdate, StateRoot) {
-        let mut state_update = tries.new_trie_update(shard_id, MerkleHash::default());
+        validators: &'a [(AccountId, PublicKey, Balance)],
+        config: &'a RuntimeConfig,
+    ) -> Self {
+        Self {
+            tries,
+            shard_id,
+            validators,
+            config,
+            current_state_root: MerkleHash::default(),
+            delayed_receipts_indices: DelayedReceiptIndices::default(),
+        }
+    }
+
+    fn commit(&mut self, mut state_update: TrieUpdate) {
+        state_update.commit(StateChangeCause::InitialState);
+        let trie_changes = state_update.finalize_genesis().expect("Genesis state update failed");
+
+        let (store_update, new_state_root) = self
+            .tries
+            .apply_all(&trie_changes, self.shard_id)
+            .expect("Failed to apply genesis chunk");
+        store_update.commit().expect("Store update failed on genesis initialization");
+        self.current_state_root = new_state_root;
+    }
+
+    fn apply_chunk(&mut self, genesis: &Genesis, chunk_account_ids: HashSet<&AccountId>) {
```

Comment:
Nomenclature: `chunk` could be confusing because in the context of nearcore, it usually refers to a shard block. I think it is better to use another name here.

Reply:
Changed it to batch